File: datasets/Part 2 - Regression/simple_linear_regression.R

# Regresion Lineal Simple

setwd("~/repos/machinelearning-az/datasets/Part 2 - Regression")

# Importar el dataset
dataset = read.csv('data/Salary_Data.csv')
#dataset = dataset[, 2:3]

# Dividir los datos en conjunto de entrenamiento y conjunto de test
library(caTools) # Instalar con install.packages('caTools')
set.seed(123) # Para que salgan los mismos resultados que en el curso
# 20 individuos para entrenar y 10 individuos para validar
split = sample.split(dataset$Salary, SplitRatio = 2/3)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)


# Escalado de valores
# Hay que indicar las columnas donde hacer el escalado
# No se hacen sobre los valores dummy de las columnas Country y Purchased porque
# no puede operar con factores que convirtio previamente a tipo string
#training_set[, 2:3] = scale(training_set[, 2:3])
#testing_set[, 2:3] = scale(testing_set[, 2:3])

# Ajustar el modelo de regresion lineal simple con el conjunto de entrenamiento
regressor = lm(formula = Salary ~ YearsExperience,
               data = training_set)

# Predecir resultados con el conjunto de test
y_pred = predict(regressor, newdata = testing_set)

# Visualizacion de los resultados en el conjunto de entrenamiento
library(ggplot2) # Instalar con install.packages('ggplot2')
ggplot() +
  geom_point(aes(x = training_set$YearsExperience,
                 y = training_set$Salary),
             colour = 'red') +
  geom_line(aes(x = training_set$YearsExperience,
                y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Sueldo vs Años de Experiencia (Conjunto de Entrenamiento)') +
  xlab('Años de Experiencia') +
  ylab('Suelo (en $)')

# Visualizacion de los resultados en el conjunto de testing
ggplot() +
  geom_point(aes(x = testing_set$YearsExperience,
                 y = testing_set$Salary),
             colour = 'red') +
  geom_line(aes(x = training_set$YearsExperience,
                y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Sueldo vs Años de Experiencia (Conjunto de Testing)') +
  xlab('Años de Experiencia') +
  ylab('Suelo (en $)')

File: datasets/Part 2 - Regression/simple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Sun May  1 11:41:11 2022

@author: mudar
"""

# Regresion Lineal Simple

# Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('data/Salary_Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 1].values


# Dividir el dataset en conjunto de entrenamiento y conjunto de testing
# En este caso se ha cogido como conjunto de test 1/3 lo cual no es lo habitual
# y recomendable pero para este caso tan simple funciona correctamente
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=0)


# En caso de regresion lineal simple no hace falta hacer escalado de variables
# porque la libreria de python que se utiliza ya lo implementa

# Escalado de variables
# Para que no haya tanta diferencia en los rangos de
# valores en los datos de las distintas columnas
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""


# Crear modelo de Regresion Lineal Simple con el conjunto de entrenamiento
from sklearn.linear_model import LinearRegression
regression = LinearRegression()
regression.fit(X_train, y_train)

# Predecir el conjunto de test
y_pred = regression.predict(X_test)

# Visualizar los resultados de entrenamiento
plt.scatter(X_train, y_train, color = 'red')
plt.plot(X_train, regression.predict(X_train), color = 'blue')
plt.title('Sueldo vs Años de Experiencia (Conjunto de Entrenamiento)')
plt.xlabel('Años de experiencia')
plt.ylabel('Sueldo (en $)')
plt.show()

# Visualizar los resultados de test
plt.scatter(X_test, y_test, color = 'red')
plt.plot(X_train, regression.predict(X_train), color = 'blue')
plt.title('Sueldo vs Años de Experiencia (Conjunto de Testing)')
plt.xlabel('Años de experiencia')
plt.ylabel('Sueldo (en $)')
plt.show()






